File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.js
function solution(m, n, board) {
  let answer = 0;
  board = board.map(v => v.split(''));
  let tmp = new Set();

  while (true) {
    for (let i = 0; i < m - 1; i++) {
      for (let j = 0; j < n - 1; j++) {
        let c = board[i][j];
        if (c === '') continue;

        if (
          board[i][j + 1] === c &&
          board[i + 1][j] === c &&
          board[i + 1][j + 1] === c
        ) {
          tmp.add(`${i},${j}`);
          tmp.add(`${i},${j + 1}`);
          tmp.add(`${i + 1},${j}`);
          tmp.add(`${i + 1},${j + 1}`);
        }
      }
    }

    if (tmp.size) {
      answer += tmp.size;
      tmp.forEach(position => {
        const [i, j] = position.split(',').map(Number);
        board[i][j] = '';
      });
      tmp.clear();
    } else break;

    while (true) {
      let moved = false;
      for (let i = m - 2; i >= 0; i--) {
        for (let j = 0; j < n; j++) {
          if (board[i][j] !== '' && board[i + 1][j] === '') {
            board[i + 1][j] = board[i][j];
            board[i][j] = '';
            moved = true;
          }
        }
      }
      if (!moved) break;
    }
  }

  return answer;
}
     

File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.py
def solution(m, n, board):
    b = [list(v) for v in board]
    answer = 0
    point = 1
    
    while point != 0:
        pop = []
        point = 0

        for i in range(m - 1):
            for j in range(n - 1):
                if b[i][j] == b[i][j + 1] == b[i + 1][j] == b[i + 1][j + 1] != '팡!':
                    pop.append((i, j))
                    point += 1

        for i, j in pop:
            b[i][j] = b[i][j + 1] = b[i + 1][j] = b[i + 1][j + 1] = '팡!'

        for j in range(n):
            for i in range(m - 1, 0, -1):
                if b[i][j] == '팡!':
                    for k in range(i - 1, -1, -1):
                        if b[k][j] != '팡!':
                            b[i][j], b[k][j] = b[k][j], '팡!'
                            break

    for v in b: answer += v.count('팡!')

    return answer